chore: update libp2p to 0.52.1

melekes opened the pull request:
changelog: https://github.com/libp2p/rust-libp2p/releases/tag/libp2p-v0.52.0

polkadot companion: https://github.com/paritytech/polkadot/pull/7472
cumulus companion: https://github.com/paritytech/cumulus/pull/2838

melekes committed:
update libp2p to 0.52.0

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3040886

melekes committed:
proto name now must implement `AsRef<str>`

melekes committed:
update libp2p version everywhere

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3042816

melekes committed:
ToSwarm, FromBehaviour, ToBehaviour

also LocalProtocolsChange and RemoteProtocolsChange

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3043865

melekes committed:
new NetworkBehaviour invariants

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3048790

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3048789

melekes committed:
replace `Vec<u8>` with `StreamProtocol`

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3054649

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3054650

melekes committed:
rename ConnectionHandlerEvent::Custom to NotifyBehaviour

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3055286

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3055285

melekes committed:
remove DialError & ListenError invariants

also fix pending_events

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3056908

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3056909

melekes committed:
use connection_limits::Behaviour

See https://github.com/libp2p/rust-libp2p/pull/3885

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3061705

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3061706

melekes committed:
impl `void::Void` for `BehaviourOut`

also use `Behaviour::with_codec`

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3075631

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3075630

melekes committed:
KademliaHandler no longer public

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3078924

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3078923

melekes committed:
fix StreamProtocol construction

melekes committed:
update libp2p-identify to 0.2.0

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3080998

melekes committed:
remove non-existing methods from PollParameters

rename ConnectionHandlerUpgrErr to StreamUpgradeError

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3083739

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3083740

melekes committed:
`P2p` now contains `PeerId`, not `Multihash`

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3083880

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3083881

melekes committed:
use multihash-codetable crate

melekes committed:
Merge branch 'master' into anton/libp2p-0.52.0

melekes committed:
update Cargo.lock

melekes committed:
reformat text

melekes committed:
comment out tests for now

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3090140

melekes commented on client/network/src/service.rs:
```diff
@@ -1556,20 +1547,28 @@ where
 					};
 					let reason = match cause {
 						Some(ConnectionError::IO(_)) => "transport-error",
+						// Some(ConnectionError::Handler(Either::Left(Either::Left(
```
no such variant seems to exist

altonen replied:
`PingFailure` doesn't exist?

What does `ConnectionError::Handler(Either::Left(Either::Left(Either::Left(Either::Right` contain?

melekes replied:
```
let reason = match cause {
     |                                        ----- this expression has type `std::option::Option<ConnectionError<either::Either<either::Either<either::Either<either::Either<NotifsHandlerError, either::Either<void::Void, std::io::Error>>, std::io::Error>, void::Void>, void::Void>>>`
```

`std::io::Error`

dmitry-markin replied:
If we are sure that this `io::Error` is handling the case for ping failure, may be we should still report it as `"ping-failure"` and not as default `"protocol-error"`?

melekes commented on client/network/src/service.rs:
```diff
@@ -1681,17 +1681,17 @@ where
-					metrics.incoming_connections_errors_total.with_label_values(&["banned"]).inc();
-				}
-			},
+			// #[allow(deprecated)]
```
TODO

melekes committed:
Merge branch 'master' into anton/libp2p-0.52.0

melekes committed:
remove `.into()` from P2p

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3091592

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3092061

melekes committed:
confirm observed addr manually

See https://github.com/libp2p/rust-libp2p/blob/master/protocols/identify/CHANGELOG.md#0430

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3117149

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3117148

melekes committed:
Merge branch 'master' into anton/libp2p-0.52.0

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3117343

melekes committed:
remove SwarmEvent::Banned

since we're not using `ban_peer_id`, this can be safely removed.
we may want to introduce `libp2p::allow_block_list` module in the future.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3117440

melekes committed:
fix imports

melekes committed:
replace `libp2p` with smaller deps in network-gossip

melekes committed:
bring back tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3118335

melekes committed:
finish rewriting tests

melekes committed:
uncomment handler tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3131383

melekes committed:
Revert "uncomment handler tests"

This reverts commit 720a06815887f4e10767c62b58864a7ec3a48e50.

melekes committed:
add a fixme

melekes commented on client/network/src/protocol/notifications/handler.rs:
```diff
@@ -1011,596 +1012,598 @@ pub mod tests {
+	// 		};
+
+	// 		handler.on_connection_event(handler::ConnectionEvent::FullyNegotiatedInbound(
+	// 			handler::FullyNegotiatedInbound { protocol: (notif_in, 0), info: () },
```
haven't found a way around this yet...

altonen replied:
The test can commented out if it's a blocker and we can look into enabling it in the future

altonen replied:
Could you create an issue so that we remember to look into this issue?

thomaseizinger replied:
I haven't looked into the details of this test but we extensively use https://docs.rs/futures_ringbuf to simulate a network connection in-memory.

altonen replied:
@thomaseizinger 

IIUC the reason these were commented out was because something in libp2p was made private. What was the exact reason @melekes?

melekes committed:
Merge branch 'master' into anton/libp2p-0.52.0

melekes committed:
update Cargo.lock

melekes committed:
remove extra From

melekes committed:
Merge branch 'master' into anton/libp2p-0.52.0

melekes commented:
looks like the node rejects both protocols (legacy & new):

```
[2023-07-12T06:06:29Z TRACE multistream_select::protocol] Received message: Protocol(Protocol("/dot/kad"))
[2023-07-12T06:06:29Z DEBUG multistream_select::listener_select] Listener: rejecting protocol: /dot/kad

[2023-07-12T06:06:29Z TRACE multistream_select::protocol] Received message: Protocol(Protocol("/0000000000000000000000000000000000000000000000000000000000000001/test-fork-id/kad"))
[2023-07-12T06:06:29Z DEBUG multistream_select::listener_select] Listener: rejecting protocol: /0000000000000000000000000000000000000000000000000000000000000001/test-fork-id/kad

[2023-07-12T06:06:29Z DEBUG libp2p_kad::behaviour] Request to PeerId("12D3KooWRrAhoqQhm83b3nTJtTdvoRU6EHGbVN1o2vNWMvBJZCtX") in query QueryId(51) failed with Upgrade(NegotiationFailed)

```

melekes commented:
@altonen maybe you have some ideas how to fix https://github.com/paritytech/substrate/pull/14429#issuecomment-1631937183 ?

altonen commented:
> @altonen maybe you have some ideas how to fix [#14429 (comment)](https://github.com/paritytech/substrate/pull/14429#issuecomment-1631937183) ?

Are you able to add debug prints to `multistream-select` and see what protocols the listener is supporting?

melekes commented:
> > @altonen maybe you have some ideas how to fix [#14429 (comment)](https://github.com/paritytech/substrate/pull/14429#issuecomment-1631937183) ?
> 
> Are you able to add debug prints to `multistream-select` and see what protocols the listener is supporting?

`yamux` and `noise`

mxinden commented:
@melekes is the rejecting node running in Kademlia `Mode::Client`?

See `v0.44.0` changelog for details: https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/CHANGELOG.md#0440

Easiest way to validate is to call [`Kademlia::set_mode`](https://docs.rs/libp2p-kad/latest/libp2p_kad/struct.Kademlia.html#method.set_mode) which was introduced in `libp2p-kad` `v0.44.2`.

``` rust
    swarm
        .behaviour_mut()
        .kademlia
        .set_mode(Some(kad::Mode::Server));
```

mxinden commented on client/authority-discovery/src/worker/addr_cache.rs:
```diff
@@ -200,14 +201,13 @@ mod tests {
-				Multihash::wrap(multihash::Code::Sha2_256.into(), &seed).unwrap(),
-			)
-			.unwrap();
+			let peer_id =
+				PeerId::from_multihash(Multihash::wrap(Code::Sha2_256.into(), &seed).unwrap())
+					.unwrap();
```
In case you don't want the explicit `multihash_codetable` dependency here, you might as well generate the `PeerId` via a `ed25519` key. For example:

``` rust
    let mut bytes = [0u8; 32];
    // Fill `bytes` via `seed`.

    identity::Keypair::ed25519_from_bytes(bytes).expect("only errors on wrong length").public().peer_id();
```

melekes replied:
I think it's fine given we are using it in multiple places, not just here

mxinden commented on client/network/src/behaviour.rs:
```diff
@@ -51,6 +51,8 @@ pub struct Behaviour<B: BlockT> {
 	peer_info: peer_info::PeerInfoBehaviour,
 	/// Discovers nodes of the network.
 	discovery: DiscoveryBehaviour,
+	/// Connection limits.
+	connection_limits: libp2p::connection_limits::Behaviour,
```
:rocket:

mxinden commented on client/network/src/behaviour.rs:
```diff
@@ -351,3 +355,9 @@ impl From<DiscoveryOut> for BehaviourOut {
 	}
 }
+
+impl From<void::Void> for BehaviourOut {
+	fn from(_e: void::Void) -> Self {
+		Self::None
+	}
+}
```
```suggestion
impl From<void::Void> for BehaviourOut {
	fn from(e: void::Void) -> Self {
		void::unreachable(e)
	}
}
```

Why not make use of the impossibility of `Void`, i.e. make use of the fact that `Void` is  _uninhabited_?

mxinden commented on client/network/src/config.rs:
```diff
@@ -181,8 +180,6 @@ impl TryFrom<String> for MultiaddrWithPeerId {
 pub enum ParseErr {
 	/// Error while parsing the multiaddress.
 	MultiaddrParse(multiaddr::Error),
-	/// Multihash of the peer ID is invalid.
-	InvalidPeerId,
```
Great to see this error case removed from runtime, but instead made impossible at compile time.

mxinden reviewed (commented):
Unsolicited review. I hope you don't mind.

melekes committed:
make void uninhabited

melekes committed:
fix discovery test

melekes committed:
use autonat protocols

confirming external addresses manually is unsafe in open networks

melekes committed:
fix SyncNotificationsClogged invariant

melekes committed:
Merge branch 'master' into anton/libp2p-0.52.0

melekes committed:
only set server mode manually in tests

doubt that we need to set it on node since we're adding public addresses

melekes commented:
> @melekes is the rejecting node running in Kademlia `Mode::Client`?
> 
> See `v0.44.0` changelog for details: https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/CHANGELOG.md#0440
> 
> Easiest way to validate is to call [`Kademlia::set_mode`](https://docs.rs/libp2p-kad/latest/libp2p_kad/struct.Kademlia.html#method.set_mode) which was introduced in `libp2p-kad` `v0.44.2`.
> 
> ```rust
>     swarm
>         .behaviour_mut()
>         .kademlia
>         .set_mode(Some(kad::Mode::Server));
> ```



> @melekes is the rejecting node running in Kademlia `Mode::Client`?
> 
> See `v0.44.0` changelog for details: https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/CHANGELOG.md#0440
> 
> Easiest way to validate is to call [`Kademlia::set_mode`](https://docs.rs/libp2p-kad/latest/libp2p_kad/struct.Kademlia.html#method.set_mode) which was introduced in `libp2p-kad` `v0.44.2`.
> 
> ```rust
>     swarm
>         .behaviour_mut()
>         .kademlia
>         .set_mode(Some(kad::Mode::Server));
> ```

ok, that explains everything. thanks 🙏

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3198707

michalkucharczyk commented on client/network/src/protocol/notifications/upgrade/notifications.rs:
```diff
@@ -203,13 +196,13 @@ impl<TSubstream> NotificationsInSubstream<TSubstream>
+	// 	handshake: NotificationsInSubstreamHandshake,
+	// ) -> Self {
+	// 	Self { socket, handshake }
+	// }
```
leftover?

melekes replied:
no, I had to comment a bunch of tests in `handler.rs`, which were using this API

michalkucharczyk commented:
codewise it looks good, however I am not familiar with all the logic.

melekes committed:
Merge branch 'master' into anton/libp2p-0.52.0

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3200675

melekes commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into anton/libp2p-0.52.0

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3202464

dmitry-markin commented on client/network/src/service.rs:
```diff
@@ -1534,6 +1527,13 @@ where
 
 				self.event_streams.send(Event::Dht(event));
 			},
+			SwarmEvent::Behaviour(BehaviourOut::AutoNAT(event)) => {
```
Is `AutoNAT` protocol the one responsible for `ExternalAddressConfirmed` events?

melekes replied:
Yes.

dmitry-markin commented on client/network/src/service.rs:
```diff
@@ -1659,21 +1664,26 @@ where
-					"Libp2p => IncomingConnectionError({},{}): {}",
-					local_addr, send_back_addr, error,
+					"Libp2p => IncomingConnectionError({},{} via {:?}): {}",
+					local_addr, send_back_addr,	connection_id, error,
```
```suggestion
					local_addr, send_back_addr, connection_id, error,
```

tomaka commented on client/network/src/behaviour.rs:
```diff
@@ -51,8 +51,12 @@ pub struct Behaviour<B: BlockT> {
+	connection_limits: libp2p::connection_limits::Behaviour,
 	/// Generic request-response protocols.
 	request_responses: request_responses::RequestResponsesBehaviour,
+	// AutoNAT.
+	auto_nat: autonat::Behaviour,
```
Is this intended? It modifies the networking protocols of Substrate/Polkadot. If yes, this should IMO go through an RFC: https://github.com/polkadot-fellows/RFCs/pulls

tomaka replied:
I find it a bit worrisome that modifying the networking protocol would have passed review

altonen replied:
Could you @melekes elaborate on why this is added in the first place?

melekes replied:
Thanks 🙏 Will create an RFC today. I've reverted the code to manually confirming the observed address (although it's not safe I believe)

melekes replied:
1. Kademlia switches its mode to server once it receives a confirmed external address. Of course, we can set mode to server ourselves (via `set_mode` like it is done in tests), but I don't think it's desirable.
2. I've moved logging of new external addresses (`Discovered new external address for our node`) to `FromSwarm::ExternalAddrConfirmed` case, so we don't log all the candidates.
3. AutoNAT makes sure external address are publicly reachable (currently we're confirming all external addresses https://github.com/paritytech/substrate/pull/14429/commits/d464408c81a83bf355758bc9c128cb12626b9ff0)

tomaka replied:
> (although it's not safe, I believe)

It's how it works today.
There's no safety aspect here. We must already consider every peer as potentially malicious. Increasing the chance that peers report accurate information should neither increase not decrease safety. If it does, it means that we don't actually consider every peer as potentially malicious and this should be fixed.

I'm not necessarily against the principle of AutoNAT. What I'm completely against is sneaking changes to the networking protocol in minor version bumps, and I'm quite annoyed that rust-libp2p does that.

mxinden replied:
Note from the rust-libp2p side:

- running `libp2p-autonat` is not required.
- `libp2p-autonat` is not automatically enabled.
- manually confirming the observed addresses is just fine. If I recall correctly there was a command line flag in Substrate where a user could even specify the external address.

Long story short, Substrate should be able to upgrade to `libp2p` `v0.52.0` without the introduction of `libp2p-autonat`.

tomaka commented:
While this is not my problem anymore, I find it frustrating that every update of a libp2p version seems to require a `+1000 -1000` pull request that understandably takes a long time to write, review, and merge.

I find it a bit ridiculous that [a very critical feature](https://github.com/paritytech/substrate/pull/12529) is blocked on [a simple version bump](https://github.com/libp2p/rust-libp2p/pull/4099), but this simple version bump comes bundles with tons of breaking changes.

melekes committed:
Merge branch 'master' into anton/libp2p-0.52.0

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3207654

altonen commented:
We all want WebRTC merged asap but this is the reality we find ourselves in and I don't think anybody likes these PRs too much. Do you have a suggestion?

melekes committed:
address @dmitry-markin comments

melekes committed:
remove autonat

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3208532

melekes committed:
removed unused var

melekes committed:
fix EOL

melekes commented on client/network/src/discovery.rs:
```diff
@@ -1023,6 +1044,16 @@ mod tests {
 					TokioExecutor(runtime),
 				)
 				.build();
+				// Set the Kademlia mode to server so that it can accept incoming requests.
+				//
+				// Note: the server mode is set automatically when the node learns its external
+				// address, but that does not happen in tests => hence we set it manually.
```
hmm.. 🤔 the discovery test fails if one node is in a client mode

altonen replied:
How about another test with one server and one client, just to verify that such a scenario also works?

melekes replied:
> How about another test with one server and one client, just to verify that such a scenario also works?

that scenario doesn't work, at least not in tests.

melekes committed:
Merge branch 'master' into anton/libp2p-0.52.0

melekes committed:
update smallvec and sha2

in attempt to compile polkadot

melekes commented:
```
[2023-07-19 08:49:06] [2023-07-19T08:49:06Z INFO  diener::patch] Adding patch for `polkadot`.
[2023-07-19 08:49:08] error: failed to select a version for `sha2`.
[2023-07-19 08:49:08]     ... required by package `sp-core-hashing v9.0.0 (/builds/parity/mirrors/substrate/primitives/core/hashing)`
[2023-07-19 08:49:08]     ... which satisfies path dependency `sp-core-hashing` of package `pallet-alliance v4.0.0-dev (/builds/parity/mirrors/substrate/frame/alliance)`
[2023-07-19 08:49:08]     ... which satisfies git dependency `pallet-alliance` (locked to 4.0.0-dev) of package `collectives-polkadot-runtime v1.0.0 (/builds/parity/mirrors/substrate/companions/cumulus/parachains/runtimes/collectives/collectives-polkadot)`
[2023-07-19 08:49:08]     ... which satisfies path dependency `collectives-polkadot-runtime` (locked to 1.0.0) of package `collectives-polkadot-integration-tests v0.1.0 (/builds/parity/mirrors/substrate/companions/cumulus/parachains/integration-tests/emulated/collectives/collectives-polkadot)`
[2023-07-19 08:49:08] versions that meet the requirements `^0.10.7` are: 0.10.7
[2023-07-19 08:49:08] 
[2023-07-19 08:49:08] all possible versions conflict with previously selected packages.
[2023-07-19 08:49:08] 
[2023-07-19 08:49:08]   previously selected package `sha2 v0.10.6`
[2023-07-19 08:49:08]     ... which satisfies dependency `sha2 = "^0.10"` (locked to 0.10.6) of package `k256 v0.13.1`
[2023-07-19 08:49:08]     ... which satisfies dependency `k256 = "^0.13.0"` (locked to 0.13.1) of package `frame-support v4.0.0-dev (/builds/parity/mirrors/substrate/frame/support)`
[2023-07-19 08:49:08]     ... which satisfies git dependency `frame-support` (locked to 4.0.0-dev) of package `asset-hub-kusama-integration-tests v1.0.0 (/builds/parity/mirrors/substrate/companions/cumulus/parachains/integration-tests/emulated/assets/asset-hub-kusama)`
```

cumulus companion fails w/ ^

melekes committed:
bump k256

in attempt to build cumulus

thomaseizinger commented:
> While this is not my problem anymore, I find it frustrating that every update of a libp2p version seems to require a `+1000 -1000` pull request that understandably takes a long time to write, review, and merge.
> 
> I find it a bit ridiculous that [a very critical feature](https://github.com/paritytech/substrate/pull/12529) is blocked on [a simple version bump](https://github.com/libp2p/rust-libp2p/pull/4099), but this simple version bump comes bundles with tons of breaking changes.

You will appreciate then that lots of engineering hours have already gone and will continue to go into making as few breaking changes as possible:

-https://github.com/multiformats/rust-multihash/issues/259
- https://github.com/multiformats/rust-multiaddr/issues/71
- https://github.com/libp2p/rust-libp2p/issues/3271
- https://github.com/libp2p/rust-libp2p/discussions/3532

You can discover more things by following the links in the above issues. Esp. the 0.52 release hardened many many APIs in rust-libp2p that allow us to make improvements without making more breaking changes :)

kpp commented on client/network/src/discovery.rs:
```diff
@@ -629,20 +629,33 @@ impl NetworkBehaviour for DiscoveryBehaviour {
+				self.kademlia.on_swarm_event(FromSwarm::NewListenAddr(e));
+
+				if let Some(ref mut mdns) = self.mdns {
+					mdns.on_swarm_event(FromSwarm::NewListenAddr(e));
```
We can use [libp2p_swarm](https://docs.rs/libp2p-swarm/latest/libp2p_swarm/index.html)::[behaviour](https://docs.rs/libp2p-swarm/latest/libp2p_swarm/behaviour/index.html)::[toggle](https://docs.rs/libp2p-swarm/latest/libp2p_swarm/behaviour/toggle/index.html)::[Toggle](https://docs.rs/libp2p-swarm/latest/libp2p_swarm/behaviour/toggle/struct.Toggle.html#) instead of `Option<TokioMdns>`. It will remove ifs because they are hidden inside `Toggle::on_*` methods.

melekes replied:
Good idea 👍 Will do in a separate PR

melekes committed:
Merge branch 'master' into anton/libp2p-0.52.0

melekes commented:
bot merge

melekes commented:
bot merge

melekes commented:
bot merge
